File: py/0121_BestTimetoBuyandSellStock.py
# 121. Best Time to Buy and Sell Stock

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        currMin = prices[0]  # min so far
        profit = 0  # max so far
        for i in range(1, len(prices)):
            currMin = min(currMin, prices[i])
            profit = max(profit, prices[i] - currMin)
        return profit


if __name__ == "__main__":
    puzzles = [
        [7, 1, 5, 3, 6, 4],
        [7, 6, 4, 3, 1],
        [3, 2, 6, 5, 0, 3],
        [3, 3, 1, 3, 3]
    ]
    for puzzle in puzzles:
        print(Solution().maxProfit(puzzle))

"""
Runtime
804
ms
Beats
12.99%
of users with Python3
Memory
27.42
MB
Beats
34.85%
of users with Python3
11
"""

File: go/0121/0121_BestTimetoBuyandSellStock.go
// 121. Best Time to Buy and Sell Stock

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"

	"github.com/samber/lo"
)

func maxProfit(prices []int) int {
	/*
		we can achieve the highest profit with the lowest price, so we should keep track of this value as we iterate
		using the lowest price, then we can determine if selling at the current iteration fetches a better price
	*/
	result := 0
	lowest := prices[0]
	for _, curr := range prices {
		lowest = min(lowest, curr)
		result = max(result, curr-lowest)
	}
	return result
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	const testSize int = 1

	for _, test := range lo.Chunk(lines, testSize) {
		var prices []int
		if err := json.Unmarshal([]byte(test[0]), &prices); err != nil {
			panic(err)
		}
		result := maxProfit(prices)
		fmt.Println(result)
	}
}
